daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Equality.scala

```diff
@@ -16,60 +12,81 @@ private[lf] object Equality {
-      h1: Iterator[X],
-      h2: Iterator[Y],
-      stack: FrontStack[(X, Y)],
-  ): FrontStack[(X, Y)] =
-    ((h1 zip h2) foldLeft stack)(_.+:(_))
+    var success = true
+    val stack = new Stack
+    stack.push(x, y)
 
-  @tailrec
-  private[this] def equality(stack0: FrontStack[(SValue, SValue)]): Boolean = {
-    stack0.pop match {
-      case Some((tuple, stack)) =>
-        tuple match {
-          case (x: SPrimLit, y: SPrimLit) =>
-            x == y && equality(stack)
-          case (SEnum(tyCon1, _, rank1), SEnum(tyCon2, _, rank2)) =>
-            tyCon1 == tyCon2 && rank1 == rank2 && equality(stack)
-          case (SRecord(tyCon1, fields1, args1), SRecord(tyCon2, fields2, args2)) =>
-            tyCon1 == tyCon2 && (fields1 sameElements fields2) &&
-              equality(zipAndPush(args1.iterator().asScala, args2.iterator().asScala, stack))
-          case (SVariant(tyCon1, _, rank1, arg1), SVariant(tyCon2, _, rank2, arg2)) =>
-            tyCon1 == tyCon2 && rank1 == rank2 && equality((arg1, arg2) +: stack)
-          case (SList(lst1), SList(lst2)) =>
-            lst1.length == lst2.length &&
-              equality(zipAndPush(lst1.iterator, lst2.iterator, stack))
-          case (SOptional(None), SOptional(None)) =>
-            equality(stack)
-          case (SOptional(Some(v1)), SOptional(Some(v2))) =>
-            equality((v1, v2) +: stack)
-          case (STextMap(map1), STextMap(map2)) =>
-            map1.keySet == map2.keySet && {
-              val keys = map1.keys
-              equality(zipAndPush(keys.iterator.map(map1), keys.iterator.map(map2), stack))
-            }
-          case (SGenMap(map1), SGenMap(map2)) =>
-            map1.keySet == map2.keySet && {
-              val keys = map1.keys
-              equality(zipAndPush(keys.iterator.map(map1), keys.iterator.map(map2), stack))
-            }
-          case (SStruct(fields1, args1), SStruct(fields2, args2)) =>
-            (fields1 sameElements fields2) && equality(
-              zipAndPush(args1.iterator().asScala, args2.iterator().asScala, stack),
-            )
-          case (SAny(t1, v1), SAny(t2, v2)) =>
-            t1 == t2 && equality((v1, v2) +: stack)
-          case (STypeRep(t1), STypeRep(t2)) =>
-            t1 == t2 && equality(stack)
-          case _ =>
-            false
-        }
-      case _ =>
+    while (success && stack.nonEmpty) success = stack.pop() match {
+      case (x: SPrimLit, y: SPrimLit) =>
+        x == y
+      case (SEnum(_, _, rank1), SEnum(_, _, rank2)) =>
+        rank1 == rank2
+      case (SRecord(_, _, args1), SRecord(_, _, args2)) =>
+        stack.push(args1.iterator().asScala, args2.iterator().asScala)
         true
+      case (SVariant(_, _, rank1, arg1), SVariant(_, _, rank2, arg2)) =>
+        stack.push(arg1, arg2)
+        rank1 == rank2
+      case (SList(lst1), SList(lst2)) =>
+        stack.push(lst1.iterator, lst2.iterator)
+        lst1.length == lst2.length
```

Comment:
This is against the spec, isn't it? Say we had a list containing one function and a list containing two functions. This code would say they are different but the spec says this should raise an exception. We probably need to allow iterators of different lengths in the `Stack` class below.